src/economy/economy.cpp

```diff
@@ -405,63 +1068,171 @@ bool nation_is_constructing_factories(sys::state& state, dcon::nation_id n) {
 	auto rng = state.world.nation_get_state_building_construction(n);
 	return rng.begin() != rng.end();
 }
-bool nation_has_closed_factories(sys::state& state, dcon::nation_id n) { // TODO - should be "good" now
-	auto nation_fat = dcon::fatten(state.world, n);
-	for(auto prov_owner : nation_fat.get_province_ownership()) {
-		auto prov = prov_owner.get_province();
-		for(auto factloc : prov.get_factory_location()) {
-			auto scale = factloc.get_factory().get_production_scale();
-			if(scale < factory_closed_threshold) {
-				return true;
-			}
-		}
-	}
-	return false;
+bool nation_has_closed_factories(sys::state& state, dcon::nation_id n) { // TODO - should be "good" now
+	auto nation_fat = dcon::fatten(state.world, n);
+	for(auto prov_owner : nation_fat.get_province_ownership()) {
+		auto prov = prov_owner.get_province();
+		for(auto factloc : prov.get_factory_location()) {
+			auto scale = factloc.get_factory().get_production_scale();
+			if(scale < factory_closed_threshold) {
+				return true;
+			}
+		}
+	}
+	return false;
+}
+
+template<typename T>
+ve::fp_vector artisan_input_multiplier(
+	sys::state& state,
+	T nations
+) {
+	auto alice_input_base = state.defines.alice_inputs_base_factor_artisans;
+	auto nation_input_mod = 1.f + state.world.nation_get_modifier_values(
+		nations, sys::national_mod_offsets::artisan_input);
+
+	return ve::max(alice_input_base * nation_input_mod, 0.01f);
+}
+
+template<typename T>
+ve::fp_vector artisan_output_multiplier(
+	sys::state& state,
+	T nations
+) {
+	auto alice_output_base = state.defines.alice_output_base_factor_artisans;
+	auto nation_output_mod = 1.f + state.world.nation_get_modifier_values(
+		nations, sys::national_mod_offsets::artisan_input);
+
+	return ve::max(alice_output_base * nation_output_mod, 0.01f);
+}
+
+template<typename T>
+ve::fp_vector artisan_throughput_multiplier(
```

Comment:
should be able to eliminate the duplicate one-item versions by making the return type `auto`